src/main/java/com/google/sps/model/GmailResponse.java

```diff
@@ -0,0 +1,288 @@
+    return (int)
+        unreadEmails.stream()
+            .filter(
+                (message) ->
+                    message.getLabelIds() != null && message.getLabelIds().contains("IMPORTANT"))
+            .count();
+  }
+
+  /**
+   * Get the sender of the most unread emails from the last nDays days in a user's Gmail account
+   *
+   * @param unreadEmails a list of unread emails (with METADATA OR FULL MessageFormat)
+   * @return sender of the most unread emails from the last nDays days
+   * @throws GmailMessageFormatException if the messages do not contain an internal date and/or
+   *     headers (and are thus the wrong format)
+   */
+  private String findMostFrequentSender(List<Message> unreadEmails)
+      throws GmailMessageFormatException {
+    // This data structure is used in the hashmap, to associate two values (how many times
+    // the sender sent emails in the last nDays, and when their most recent sent email was) with
+    // the sender's email in the hashmap.
+    class EmailFrequencyWithDate {
+      private int timesSent;
+      private long mostRecentEmailTimestamp;
+
+      /**
+       * Create a new EmailFrequencyWithDate instance
+       *
+       * @param timestamp timestamp of email sent in milliseconds since epoch
+       */
+      public EmailFrequencyWithDate(long timestamp) {
+        timesSent = 1;
+        this.mostRecentEmailTimestamp = timestamp;
+      }
+
+      public int getTimesSent() {
+        return timesSent;
+      }
+
+      public long getMostRecentEmailTimestamp() {
+        return mostRecentEmailTimestamp;
+      }
+
+      /**
+       * Updates the mostRecentEmailTimestamp if the passed timestamp is later than the currently
+       * stored one.
+       *
+       * @param timestamp timestamp of email sent in milliseconds since epoch
+       */
+      public void updateMostRecentEmailTimestamp(long timestamp) {
+        if (timestamp > mostRecentEmailTimestamp) {
+          this.mostRecentEmailTimestamp = timestamp;
+        }
+      }
+
+      public void incrementTimesSent() {
+        timesSent++;
+      }
+    }
+
+    // If there are no emails, return "" (the default value for sender)
+    if (unreadEmails.isEmpty()) {
+      return "";
+    }
+
+    // Find out which sender sent the most unread emails.
+    // This is extracted from the "From" headers within messages (which are always present in a
+    // valid email)
+    // Store the senders of emails in a hashmap paired with 1) how often they sent emails in the
+    // last nDays and 2) what the timestamp was of their most recent email.
+    HashMap<String, EmailFrequencyWithDate> senders = new HashMap<>();
+    unreadEmails.forEach(
+        (Message m) -> {
+          long timestamp = m.getInternalDate();
+          MessagePart messagePayload = m.getPayload();
+
+          if (messagePayload == null || timestamp == 0) {
+            throw new GmailMessageFormatException("Messages must be of format METADATA or FULL");
+          }
```

Comment:
Can we perform a sanity check on `unreadEmails` only once? Maybe in the constructor to avoid having multiple checks throughout your auxiliary methods.

Reply:
I agree with you for this case. However, now that I've moved a lot of the logic into utility methods that are standalone, I think having multiple copies of the checks is important in case one of the new utility methods is used on their own. 

Let me know what you think with the new changes